text/0000-custom-dst.md

```diff
@@ -0,0 +1,480 @@
+- Feature Name: `custom_dst`
+- Start Date: 2018-11-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow Rust code to define dynamically sized types with custom thick (and thin)
+pointers, and define slice functions in terms of these, instead of transmute.
+Also, convert the `CStr` type to use this functionality,
+and make it a thin pointer; this will allow use with FFI.
+
+# Motivation
+[motivation]: #motivation
+
+As of right now, the lack of custom DSTs in Rust means that we can't communicate
+with C in important ways - we lack the ability to define a `CStr` in the
+language such that `&CStr` is compatible with `char const *`,
+and we lack the ability to communicate nicely with C code that uses
+[Flexible Array Members](https://en.wikipedia.org/wiki/Flexible_array_member).
+This RFC attempts to fix this,
+as well as introduce more correctness to existing practices.
+
+Apart from FFI, it also has usecases for indexing and slicing 2-d arrays.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+There are two new language traits in the standard library, similar to `Sized`,
+under `std::marker`:
+
+```rust
+// All types satisfy the constraint <T: Pointee>
+trait Pointee {
+    // note: required for
+    //  - Unpin + Copy - all pointer types
+    //  - Send + Sync - &T, &mut T
+    //  - Eq + Ord - *const T, *mut T
+    type Metadata: 'static + Copy + Send + Sync + Eq + Ord + Unpin;
```

Comment:
When written this way, these bounds can never be relaxed later as doing so would be a breaking change. An alternative formulation would be to turn them into constraints.

```rust
trait Pointee
    where <Self as Pointee>::Metadata: 'static + Copy + Send + Sync + Eq + Ord + Unpin,
{
    type Metadata;
}
```

It's probably not wise most beneficial to move _all_ bounds. But can we be sure all future use cases for extern types guarantee `'static` and `Send`, `Sync`, `Unpin` in particular?

Reply:
> An alternative formulation would be to turn them into constraints.

Both are handled almost identically in the compiler. In either case a user of `Pointee` is allowed to depend on it.

Reply:
Hm, I didn't realize. Then what would be a proper way of introducing those constraints?

Reply:
I don't think it is possible to add constraints to the implementer without allowing the user to depend on the constraints.

Reply:
Is it realistic that these bounds can be relaxed?

* `'static` - even something as simple as creating `Box<TypeWithNonStaticMetadata>` and calling `Box::leak` would invalidate it
* `Copy` - creating shared reference would become impossible (yeah, you didn't mention it, but for completeness...)
* `T: Sync` implies `&T: Send` and the users can rely on it today. `print_in_other_thread` would stop compiling in this [example](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=e3e1078ed138caf9904636b6c5c54f8d) if the bounds were relaxed
* Relaxing `Unpin` would mean that it'd be unsound to move `&mut T`. Even if conceivable, it'd be a massive footgun.

Reply:
> 'static - even something as simple as creating Box<TypeWithNonStaticMetadata> and calling Box::leak would invalidate it

Who's to say that all custom DSTs can be boxed?

> T: Sync implies &T: Send and the users can rely on it today. print_in_other_thread would stop compiling in this example if the bounds were relaxed

As noted `T: ?Sized` will imply `T: Contiguous` and similarly could imply `<T as Pointee>::Metadata: Sync` etc. such that all soundness requirements of the existing `impl<T: ?Sized + Sync> Send for &T` are fulfilled. Then that `print_in_other_thread` would need to be relaxed further when it wants to support arbitrary metdata.

> Relaxing Unpin would mean that it'd be unsound to move &mut T. Even if conceivable, it'd be a massive footgun.

Same as above, if `Unpin` is implied by `T: ?Sized` (or stronger) then the footgun is hardly massive.

Reply:
Yeah, that could work, I think.

> Then that `print_in_other_thread` would need to be relaxed further when it wants to support arbitrary metdata.

It's a bit sad that many generics will have to be modified to become more flexible but there's probably no better solution. At least I don't know of such. Perhaps an automated analysis tool would help here.